FaceDetectionServer/moedx/object_detection/object_detector.py

```diff
@@ -0,0 +1,189 @@
+
+package_dir = os.path.dirname(os.path.abspath(__file__))+"/"+submodule
+config_path=package_dir+'/config/yolov3.cfg'
+weights_path=package_dir+'/config/yolov3.weights'
+class_path=package_dir+'/config/coco.names'
+img_size=416
+conf_thres=0.8
+nms_thres=0.4
+
+# Load model and weights
+model = Darknet(config_path, img_size=img_size)
+model.load_weights(weights_path)
+
+if torch.cuda.is_available():
+    model.cuda()
+    Tensor = torch.cuda.FloatTensor
+else:
+    Tensor = torch.FloatTensor
+
+model.eval()
+classes = utils.load_classes(class_path)
+
+# Globals
+total_server_processing_time = 0
+count_server_processing_time = 0
+
+class ObjectDetector(object):
+    def __init__(self):
+        pass
+
+    def detect_image(self, img):
+        # scale and pad image
+        ratio = min(img_size/img.size[0], img_size/img.size[1])
+        imw = round(img.size[0] * ratio)
+        imh = round(img.size[1] * ratio)
+        img_transforms = transforms.Compose([ transforms.Resize((imh, imw)),
+             transforms.Pad((max(int((imh-imw)/2),0), max(int((imw-imh)/2),0), max(int((imh-imw)/2),0), max(int((imw-imh)/2),0)),
+                            (128,128,128)),
+             transforms.ToTensor(),
+             ])
+        # convert image to Tensor
+        image_tensor = img_transforms(img).float()
+        image_tensor = image_tensor.unsqueeze_(0)
+        input_img = Variable(image_tensor.type(Tensor))
+        # run inference on the model and get detections
+        with torch.no_grad():
+            detections = model(input_img)
+            detections = utils.non_max_suppression(detections, 80, conf_thres, nms_thres)
+        return detections[0]
+
+    def process_image(self, img, outdir = None):
+        """
+        Convert and process the image. If "outdir" is included,
+        save the processed image there. (Only for command-line executions.
+        It will always be None when called by the Django server.)
+        """
+        global total_server_processing_time
+        global count_server_processing_time
+        # load image and get detections
+        prev_time = time.time()
+        detections = self.detect_image(img)
+        millis = (time.time() - prev_time)*1000
+        elapsed = "%.3f" %millis
+        print("%s ms to detect objects" %(elapsed))
+        count_server_processing_time += 1
+        total_server_processing_time += millis
+
+        if outdir is not None:
+            filename = img.filename
+            img = np.array(img)
+            # Get bounding-box colors
+            cmap = plt.get_cmap('tab20b')
+            colors = [cmap(i) for i in np.linspace(0, 1, 20)]
+
+            plt.figure()
+            fig, ax = plt.subplots(1, figsize=(12,9))
+            ax.imshow(img)
+        else:
+            img = np.array(img)
+
+        pad_x = max(img.shape[0] - img.shape[1], 0) * (img_size / max(img.shape))
+        pad_y = max(img.shape[1] - img.shape[0], 0) * (img_size / max(img.shape))
```

Comment:
I'm a little bit confused by the padding logic (also in line 69). Are we subtracting the height from the width and vice versa? Is this to make the image a square?

Reply:
Yes, the image must be a square, but with gray padding to maintain aspect ration. Picture a letterboxed or pillarboxed movie.